File: 07-trees/00105-Construct_Binary_Tree_From_Preorder_And_Inorder_Traversal/00105-construct_binary_tree_from_preorder_and_inorder_traversal.java
// 105: Construct Binary Tree From Preorder And Inorder Traversal
// https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal



import java.util.Queue;
import java.util.ArrayList;
import java.util.LinkedList;


class Solution {
    public static class TreeNode {
        Integer val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(Integer val) { this.val = val; }
        TreeNode(Integer val, TreeNode left, TreeNode right) { this.val = val; this.left = left; this.right = right; }
    }

    public static class Tree {
        TreeNode root;
        Tree() { this.root = null; }
    }

    private static TreeNode creator(Integer[] values, TreeNode root, int i, int n) {
        if (n==0) return null;
        if (i<n) {
            TreeNode temp = new TreeNode(values[i]);
            root = temp;
            root.left = creator(values, root.left, 2*i+1, n);
            root.right = creator(values, root.right, 2*i+2, n);
        }
        return root;
    }

    public static void createTree(Tree tree, Integer[] inputs) {
        tree.root = creator(inputs, tree.root, 0, inputs.length);
    }

    public static void showTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        ArrayList<ArrayList<Integer>> result = new ArrayList<>();
        ArrayList<Integer> c = new ArrayList<>();
        if (root == null) { System.out.println("Empty !"); return; }
        q.add(root);
        q.add(null);
        while (!q.isEmpty()) {
            TreeNode t = q.poll();
            if (t==null) {
                result.add(c);
                c = new ArrayList<>();
                if (q.size() > 0) q.add(null);
            } else {
                c.add(t.val);
                if (t.left!=null) q.add(t.left);
                if (t.right!=null) q.add(t.right);
            }
        }

        System.out.print("["); for (var x : result) {
            System.out.print("["); for (var y : x) {
                if (y==null) { System.out.print("NULL,"); continue; }
                System.out.print(y+",");
            } System.out.print("\b],");
        } System.out.println("\b]");
    }


    // SOLUTION
    private TreeNode build(int preStart, int inStart, int inEnd, Integer[] preorder, Integer[] inorder) {
        if (preStart > preorder.length - 1 || inStart > inEnd) return null;

        TreeNode root = new TreeNode(preorder[preStart]);

        int inIndex = 0;
        for (int i = inStart; i <= inEnd; i++)
            if (inorder[i] == root.val)
                inIndex = i;

        root.left = build(preStart + 1, inStart, inIndex - 1, preorder, inorder);
        root.right = build(preStart + inIndex - inStart + 1, inIndex + 1, inEnd, preorder, inorder);

        return root;
    }

    public TreeNode buildTree(Integer[] preorder, Integer[] inorder) {
        return build(0, 0, inorder.length - 1, preorder, inorder);
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        Integer[] preorder = {3,9,20,15,7};
        Integer[] inorder = {9,3,15,20,7};

        // OUTPUT
        var result = o.buildTree(preorder, inorder);
        showTree(result);
    }
}

File: 07-trees/00105-Construct_Binary_Tree_From_Preorder_And_Inorder_Traversal/00105-construct_binary_tree_from_preorder_and_inorder_traversal.py
# 105: Construct Binary Tree From Preorder And Inorder Traversal
# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal

from typing import List

class TreeNode:
    def __init__(self, x: int):
        self.val = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self) -> None:
        self.root = None

    def creator(self, values: list[int], root: TreeNode, i: int, n: int) -> TreeNode:
        if n==0 : return None
        if i<n:
            temp = TreeNode(values[i])
            root = temp
            root.left = self.creator(values, root.left, 2*i+1, n)
            root.right = self.creator(values, root.right, 2*i+2, n)
        return root

    def createTree(self, inputs: list[int]) -> None:
        self.root = self.creator(inputs, self.root, 0, len(inputs))
    
    def showTree(self, root: TreeNode) -> None:
        q = []
        result = [[]]
        c = []
        if root==None : return result
        q.append(root)
        q.append(None)
        while len(q)!=0:
            t = q.pop(0)
            if t==None:
                result.append(c)
                c = []
                if len(q) > 0 : q.append(None)
            else:
                c.append(t.val)
                if t.left!=None : q.append(t.left)
                if t.right!=None : q.append(t.right)

        print("[", end="")
        for x in result:
            if len(x)==0 : continue
            print("[", end="")
            for y in x:
                if y==None: 
                    print("NULL,", end="")
                    continue
                print(y, end=",")
            print("\b]", end=",")
        print("\b]")


class Solution:
    # SOLUTION
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def build(preStart: int, inStart: int, inEnd: int, preorder: List[int], inorder: List[int]) -> TreeNode:
            if preStart > len(preorder) - 1 or inStart > inEnd: return None

            root = TreeNode(preorder[preStart])

            inIndex = 0
            for i in range(inStart, inEnd+1):
                if (inorder[i] == root.val):
                    inIndex = i

            root.left = build(preStart + 1, inStart, inIndex - 1, preorder, inorder)
            root.right = build(preStart + inIndex - inStart + 1, inIndex + 1, inEnd, preorder, inorder)

            return root

        return build(0, 0, len(inorder) - 1, preorder, inorder)


if __name__ == "__main__":
    o = Solution()
    tree = Tree()

    # INPUT
    preorder = [3,9,20,15,7]
    inorder = [9,3,15,20,7]

    # OUTPUT
    result = o.buildTree(preorder, inorder)
    tree.showTree(result)
